### PR TITLE
TimeMap bug for RPTRST BASIC=4 terminated before a year!

### PR DESCRIPTION
**Pull request by joakim-hove**
Downstream: 

opm-output: https://github.com/OPM/opm-output/pull/54
~~opm-data: https://github.com/OPM/opm-data/pull/112~~

### CODE REVIEW

**Comment by joakim-hove** [comment]
@atgeirr: I hope and think this should fix the issue with `END` in the SPE9 deck.

**Comment by jokva** [comment]
I'll have a closer look at the details here tomorrow, but looks good at a glance.

File: opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 1]
```diff
@@ -58,17 +58,6 @@
 
 namespace Opm {
 
-    static inline std::shared_ptr< TimeMap > createTimeMap( const Deck& deck ) {
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: I'm happy this gets to go.

File: opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 2]
```diff
@@ -152,19 +141,14 @@ namespace Opm {
         for (size_t keywordIdx = 0; keywordIdx < section.size(); ++keywordIdx) {
             const auto& keyword = section.getKeyword(keywordIdx);
 
-            if (keyword.name() == "DATES") {
-                handleDATES(keyword);
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: `handleDATES` and `handleTSTEP` are no longer used as far as I can see, so feel free to completely remove them. Good change!

**Comment by atgeirr** [comment]
This works well now with SPE9.DATA, adding END to various places.

However a test with the Norne deck was unsuccessful, adding END at line 1361 of (in opm-data) norne/INCLUDE/BC0407_HIST01122006.SCH fails to parse:
```
Failed to create valid EclipseState object. See logfile: output_norne439/NORNE_439DAYS.PRT
Exception caught: Key not found:F-1H
```

File: opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 2 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: Yes - they should be removed.

**Comment by joakim-hove** [comment]
> This works well now with SPE9.DATA, adding END to various places.

Good

> However a test with the Norne deck was unsuccessful, adding END at line 1361 of (in opm-data) norne/INCLUDE/BC0407_HIST01122006.SCH fails to parse:

Not so good - this is a completely unrelated bug; the well `F-1H` is introduced via `WELSPECS` *after* line 1361 in the Schedule file; and I think the exception you get is in the `SummaryConfig` which fails to work with an unrecognized well name. Traceback?

**Comment by atgeirr** [comment]
Warning: some EXTREMELY long lines here:
```
    frame #0: 0x00007fff8b308ab7 libc++abi.dylib`__cxa_throw
  * frame #1: 0x0000000100b221e3 flow`Opm::OrderedMap<std::__1::shared_ptr<Opm::Well> >::get(this=0x00007fff5fbfdaf0, key="F-1H") const + 611 at OrderedMap.hpp:80
    frame #2: 0x0000000100b1a28f flow`Opm::Schedule::getWell(this=0x00007fff5fbfdae0, wellName="F-1H") const + 47 at Schedule.cpp:1338
    frame #3: 0x0000000100cc131f flow`Opm::keywordC(keyword=0x000000010285ab18, schedule=0x00007fff5fbfdae0, dims=array<int, 3> @ 0x00007fff5fbfcec0) + 5023 at SummaryConfig.cpp:207
    frame #4: 0x0000000100cbee74 flow`Opm::handleKW(keyword=0x000000010285ab18, schedule=0x00007fff5fbfdae0, props=0x00007fff5fbfdbc8, n_xyz=array<int, 3> @ 0x00007fff5fbfd930) + 804 at SummaryConfig.cpp:230
    frame #5: 0x0000000100cc2295 flow`Opm::SummaryConfig::SummaryConfig(Opm::Deck const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>) [inlined] decltype(__f=0x00007fff5fbfdad8, __args=0x000000010285ab18, __args=0x00007fff5fbfdae0, __args=0x00007fff5fbfdbc8, __args=0x00007fff5fbfdd18)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>)>(fp)(std::__1::forward<Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul> const&>(fp0))) std::__1::__invoke<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const&)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul> const&>(std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const&&&)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), Opm::DeckKeyword const&&&, Opm::Schedule const&&&, Opm::Eclipse3DProperties const&&&, std::__1::array<int, 3ul> const&&&) + 162 at __functional_base:415
    frame #6: 0x0000000100cc21f3 flow`Opm::SummaryConfig::SummaryConfig(Opm::Deck const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>) [inlined] std::__1::__bind_return<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::tuple<std::__1::placeholders::__ph<1>, Opm::Schedule, Opm::Eclipse3DProperties, std::__1::array<int, 3ul> > const, std::__1::tuple<Opm::DeckKeyword const&>, _is_valid_bind_return<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::tuple<std::__1::placeholders::__ph<1>, Opm::Schedule, Opm::Eclipse3DProperties, std::__1::array<int, 3ul> > const, std::__1::tuple<Opm::DeckKeyword const&> >::value>::type std::__1::__apply_functor<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (__f=0x00007fff5fbfdad8, __bound_args=0x00007fff5fbfdae0, (null)=__tuple_indices<0, 1, 2, 3> @ 0x00007fff5fbfe118, __args=0x00007fff5fbfe0d8)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::tuple<std::__1::placeholders::__ph<1>, Opm::Schedule, Opm::Eclipse3DProperties, std::__1::array<int, 3ul> > const, 0ul, 1ul, 2ul, 3ul, std::__1::tuple<Opm::DeckKeyword const&> >(std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const&)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::tuple<std::__1::placeholders::__ph<1>, Opm::Schedule, Opm::Eclipse3DProperties, std::__1::array<int, 3ul> > const&, std::__1::__tuple_indices<0ul, 1ul, 2ul, 3ul>, std::__1::tuple<Opm::DeckKeyword const&>&&) + 287 at functional:2060
    frame #7: 0x0000000100cc20d4 flow`Opm::SummaryConfig::SummaryConfig(Opm::Deck const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>) [inlined] std::__1::__bind_return<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::tuple<std::__1::placeholders::__ph<1>, Opm::Schedule, Opm::Eclipse3DProperties, std::__1::array<int, 3ul> > const, std::__1::tuple<Opm::DeckKeyword const&&&>, _is_valid_bind_return<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (* const)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::tuple<std::__1::placeholders::__ph<1>, Opm::Schedule, Opm::Eclipse3DProperties, std::__1::array<int, 3ul> > const, std::__1::tuple<Opm::DeckKeyword const&&&> >::value>::type std::__1::__bind<std::__1::vector<ERT::smspec_node, std::__1::allocator<ERT::smspec_node> > (this=0x00007fff5fbfdad8, __args=0x000000010285ab18)(Opm::DeckKeyword const&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>), std::__1::placeholders::__ph<1>&, Opm::Schedule const&, Opm::Eclipse3DProperties const&, std::__1::array<int, 3ul>&>::operator()<Opm::DeckKeyword const&>(Opm::DeckKeyword const&&&) const + 164 at functional:2132
    frame #8: 0x0000000100cc2030 flow`Opm::SummaryConfig::SummaryConfig(this=0x000000010a92fd40, deck=0x00000001021b19c0, schedule=0x000000010ab9cc88, props=0x000000010a92fbc0, n_xyz=array<int, 3> @ 0x00007fff5fbfdda8) + 1488 at SummaryConfig.cpp:258
    frame #9: 0x0000000100cc289d flow`Opm::SummaryConfig::SummaryConfig(this=0x000000010a92fd40, deck=0x00000001021b19c0, schedule=0x000000010ab9cc88, props=0x000000010a92fbc0, n_xyz=array<int, 3> @ 0x00007fff5fbfe4e0) + 93 at SummaryConfig.cpp:246
    frame #10: 0x0000000100aba9b7 flow`Opm::EclipseConfig::EclipseConfig(this=0x000000010a92fd10, deck=0x00000001021b19c0, eclipse3DProperties=0x000000010a92fbc0, inputGrid=0x000000010a92fb88, schedule=0x000000010ab9cc88) + 439 at EclipseConfig.cpp:42
    frame #11: 0x0000000100abac25 flow`Opm::EclipseConfig::EclipseConfig(this=0x000000010a92fd10, deck=0x00000001021b19c0, eclipse3DProperties=0x000000010a92fbc0, inputGrid=0x000000010a92fb88, schedule=0x000000010ab9cc88) + 53 at EclipseConfig.cpp:43
    frame #12: 0x0000000100aa9688 flow`Opm::EclipseState::EclipseState(this=0x000000010a92fa90, deck=0x00000001021b19c0, parseContext=<unavailable>) + 760 at EclipseState.cpp:64
    frame #13: 0x0000000100aaa7bd flow`Opm::EclipseState::EclipseState(this=0x000000010a92fa90, deck=0x00000001021b19c0, parseContext=<unavailable>) + 29 at EclipseState.cpp:68
    frame #14: 0x0000000100aaa76f flow`Opm::EclipseState::EclipseState(this=0x000000010a92fa90, deckptr=<unavailable>, parseContext=<unavailable>) + 63 at EclipseState.cpp:54
    frame #15: 0x000000010004e0f4 flow`Opm::FlowMainBase<Opm::FlowMain<UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >, UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >::readDeckInput(this=0x00007fff5fbff798) + 2468 at FlowMain.hpp:415
    frame #16: 0x0000000100014f11 flow`Opm::FlowMainBase<Opm::FlowMain<UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >, UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >::execute(this=0x00007fff5fbff798, argc=3, argv=0x00007fff5fbff928) + 433 at FlowMain.hpp:142
    frame #17: 0x000000010000a612 flow`main(argc=3, argv=0x00007fff5fbff928) + 66 at flow.cpp:41
    frame #18: 0x00007fff833db5c9 libdyld.dylib`start + 1
    frame #19: 0x00007fff833db5c9 libdyld.dylib`start + 1
```

**Comment by jokva** [comment]
Yes, and `std::bind` it seems.

**Comment by jokva** [comment]
Error is pretty much as expected - since a well is requested **by name** it is assumed that it is properly specified in the deck. Since the Deck is incomplete it throws.

Personally I'm not sure that this should be properly considered an error (why do you request a well you do not have? Broken input, which means output is likely to be garbage anyway, and I like to know that).

Anywho, that's a simple fix. It works well with an unmodified or a less broken Norne, right?

**Comment by atgeirr** [comment]
> why do you request a well you do not have

I don't. All usage of that well appears after the inserted END. It goes just as wrong if I simply delete all the contents after that.

**Comment by jokva** [comment]
You do. The summary include is on line 435 in `NORNE_ATW.DATA`, and `BC0407_HIST01122006.SCH` is included on line 644.

**Comment by atgeirr** [comment]
I see. That unfortunately makes the data file a lot harder to hack (I guess it would have worked back when we ignored the SUMMARY section...) WWED? (What would eclipse do?)

I'd like to see this generate an OpmLog::warning() instead of a throw, it is not going to change results, and can be safely ignored: the only risk is that the user might have made a typo and meant another well, and the warning should take care of that pretty well.

**Comment by joakim-hove** [comment]
I think this should be handled with a ParseContext error mode.

**Comment by OPMUSER** [comment]
Atgeirr,

Eclipse would report the error on the log and print file and continue, 
basically switching on the NOSIM option if errors occur.

ThunderBird Signature File

Regards, David Baxendale Email: David.Baxendale@gmail.com

------------------------------------------------------------------------

**Comment by joakim-hove** [comment]
>Eclipse would report the error on the log and print file and continue,
basically switching on the NOSIM option if errors occur.

Now - I am not the one to dispute this; but I must say I find it highly out-of-character that ECLIPSE switches to NOSIM; i.e. for all practical purposes failing. Anyway - then we can throw as default mohahhaa :-)

**Comment by OPMUSER** [comment]
Eclipse will not run if errors are encountered but continues to process the
input deck. I f warnings are encounter the run will continue and will
execute. So the behavior for errors, is identical to NOSIM.

Regards

David Baxendale
david.baxendale@gmail.com

**Comment by joakim-hove** [comment]
Actually the last commit in this PR; addressing the second problem mentioned by @atgeirr turned out to be largest change - by far.

**Comment by jokva** [comment]
Looks good, merge whenever from me.

**Comment by jokva** [comment]
From travis: 

> terminate called after throwing an instance of 'std::invalid_argument'
  what():  SUMMARY_UNKNOWN_WELL: Error in keyword:WOPR No such well: PRO1
Loading deck: /home/travis/build/OPM/opm-parser/../opm-data/spe10model2/SPE10_MODEL2.DATA ..... parse complete - creating EclipseState .... 

I call that a success to be honest. This seems mergable once tests are updated accordingly.

**Comment by joakim-hove** [comment]
jenkins build this with opm-output=54 please

**Comment by joakim-hove** [comment]
jenkins build this with downstreams opm-output=54 please

**Pull request merged**